### PR TITLE
Upgrade chroma to v2.3.0

### PR DESCRIPTION
**Pull request by silverwind**
The behaviour of `PreventSurroundingPre` has changed in https://github.com/alecthomas/chroma/pull/618 so that apparently it now causes line wrapper tags to be no longer emitted, but we need some form of indication to split the HTML into lines, so I did what https://github.com/yuin/goldmark-highlighting/pull/33 did and added the `nopWrapper`.

Maybe there are more elegant solutions but for some reason, just splitting the HTML string on `\n` did not work.

### CODE REVIEW

**Review by lunny — APPROVED** [review]

**Review by jolheiser — APPROVED** [review]

**Comment by silverwind** [comment]
It might be possible to use `chroma.Iterator` directly in highlight.go so we can skip the joining and splitting of the HTML altogether and `nopWrapper` wouldn't be needed as well, checking...

**Comment by silverwind** [comment]
Now it uses chroma's `SplitTokensIntoLines` instead of homebrew line splitting.

It seems to work fine, but I'm unsure whether that `for` loop is as efficient as it can get because if flushs the writer and resets the buffer every iteration. Would appreciate feedback on whether this can be optimized.

File: modules/highlight/highlight.go [hunk 1]
```diff
@@ -175,28 +172,30 @@ func File(fileName, language string, code []byte) ([]string, error) {
...
-		line = strings.TrimSuffix(line, "</span></span>")
-		m = append(m, line)
+	lines := []string{}
+	for _, tokens := range(tokensArray) {
+		iterator := chroma.Literator(tokens...)
+		err = formatter.Format(htmlWriter, styles.GitHub, iterator)
+		if err != nil {
+			return nil, fmt.Errorf("can't format code: %w", err)
+		}
+		_ = htmlWriter.Flush()
```
Comment (silverwind) [inline modules/highlight/highlight.go:192]: I wonder whether it's safe to omit this `Flush`.

File: services/gitdiff/highlightdiff_test.go [hunk 2]
```diff
@@ -21,11 +21,11 @@ func TestDiffWithHighlight(t *testing.T) {
 		"		run(db)\n",
 	)
 
-	expected := `		<span class="n">run</span><span class="o">(</span><span class="removed-code"><span class="k">&#39;</span><span class="o">&lt;</span><span class="o">&gt;</span><span class="k">&#39;</span></span><span class="o">)</span>` + "\n"
+	expected := `		<span class="n">run</span><span class="o">(</span><span class="removed-code"><span class="k">&#39;</span><span class="o">&lt;</span><span class="o">&gt;</span><span class="k">&#39;</span></span><span class="o">)</span>`
```
Comment (wxiaoguang) [inline services/gitdiff/highlightdiff_test.go:24]: Maybe you would have the problem "copy but EOL lost" again if there is no `\n` during rendering.

Comment (silverwind) [reply services/gitdiff/highlightdiff_test.go:24]: Hmm yes, need to actually test this.

Comment (silverwind) [reply services/gitdiff/highlightdiff_test.go:24]: Seems to copy fine in Firefox in diffs and in code view.

Comment (silverwind) [reply services/gitdiff/highlightdiff_test.go:24]: Chrome is also okay, so I think this change has no impact on copy-paste.

**Comment by wxiaoguang** [comment]
Some small changes:
* remove `htmlWriter` (no need to have it or do flush), the `htmlBuf.Reset` will reuse the underlying allocated memory, so no performance problem
* rename tokensArray to `tokensLines` to match the function `chroma.SplitTokensIntoLines`
* pre-allocate memory for `lines` with `make`

and LGTM

**Comment by lunny** [comment]
make L-G-T-M work

**Pull request merged**